### PR TITLE
CFE-3672: Fixed crash when attempting to put methods promises in bundles which are not agent bundles

### PR DESCRIPTION
**Pull request by KotzaBoss**
Ticket: CFE-3672
Changelog: Title

### CODE REVIEW

**Comment by KotzaBoss** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6643/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6643/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6643/

File: libpromises/policy.c [hunk 1]
```diff
@@ -582,6 +582,15 @@ static bool ConstraintCheckSyntax(const Constraint *constraint, Seq *errors)
 
     /* Check if lvalue is valid for the bundle's specific section. */
     const PromiseTypeSyntax *promise_type_syntax = PromiseTypeSyntaxGet(bundle->type, section->promise_type);
+
+    if(promise_type_syntax == NULL)
+    {
+        SeqAppend(errors,
+                  PolicyErrorNew(POLICY_ELEMENT_TYPE_CONSTRAINT, constraint,
+                                 POLICY_ERROR_CONSTRAINT_TYPE_MISMATCH, section->promise_type));
```
Comment (olehermanse) [inline libpromises/policy.c:590]: I think this will produce a very confusing error message. Please try it yourself.

**Review by olehermanse — CHANGES_REQUESTED** [review]
As discussed in slack, this should have an entry in the changelog. When you put `Changelog: None` it means it won't be included in the changelog.

**Comment by KotzaBoss** [comment]
An interesting predicament appeared.
```
bundle common segv
{
  methods:
      "some";
}
```
Raises the [this](https://github.com/cfengine/core/pull/4683/files#diff-9bb2a213157ff70eaff2a9f1cf7a7a7702c0d43cc62f415140801f542d089946R2810) assertion. I believe i tracked down why.
In [these](https://github.com/cfengine/core/pull/4683/files#diff-9bb2a213157ff70eaff2a9f1cf7a7a7702c0d43cc62f415140801f542d089946R2788-R2799) lines is_builtin is true however, the conlist has length 0 therefore it never calls the ConstraintsCheck function to catch the error of having a unsupported.
I pushed a separete commit to address this

**Comment by KotzaBoss** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6646/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6646/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6646/

File: libpromises/policy.c [hunk 2]
```diff
@@ -584,15 +584,6 @@ static bool ConstraintCheckSyntax(const Constraint *constraint, Seq *errors)
...
-
-    if(promise_type_syntax == NULL)
-    {
-        SeqAppend(errors,
-                  PolicyErrorNew(POLICY_ELEMENT_TYPE_BUNDLE_SECTION, section,
-                                 POLICY_ERROR_PROMISE_TYPE_UNSUPPORTED, section->promise_type, bundle->type));
-        return false;
-    }
-
     for (size_t i = 0; promise_type_syntax->constraints[i].lval != NULL; i++)
```
Comment (olehermanse) [inline libpromises/policy.c]: By removing the null pointer check, you've introduced a new segfault here.

**Review by olehermanse — CHANGES_REQUESTED** [review]

**Comment by olehermanse** [comment]
> bundle common segv
> {
>   methods:
>       "some";
> }



> An interesting predicament appeared.
> 
> ```
> bundle common segv
> {
>   methods:
>       "some";
> }
> ```
> 
> Raises the [this](https://github.com/cfengine/core/pull/4683/commits/97e13c4d046ac6236a9e5a07ba59a19cea954198#diff-9bb2a213157ff70eaff2a9f1cf7a7a7702c0d43cc62f415140801f542d089946R2810) assertion. I believe i tracked down why.
> In [these](https://github.com/cfengine/core/blob/master/libpromises/policy.c#L2777-L2787) lines is_builtin is true however, the conlist has length 0 therefore it never calls the ConstraintsCheck function to catch the error of having a unsupported promise type.
> I pushed a separate commit to address this

@KotzaBoss You should add this policy as a test. It can be an errorexit test `.x.cf` or errorlog test `.error.cf`, but it is definitely good to test that it doesn't segfault :)

**Review by olehermanse — CHANGES_REQUESTED** [review]
The 2 commits don't make much sense to me. It would be much easier to review and understand if you just squashed them into 1 commit. Since in the second one you undo some changes from the first one. Also, the second one says "restructure", but looking at the changes, I see more than just restructuring.

**Comment by olehermanse** [comment]
@cf-bottom please test this in jenkins, again.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6650/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6650/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6650/

**Review by olehermanse — APPROVED** [review]

**Pull request merged**